test/messenger/messenger_test.cpp

```diff
@@ -0,0 +1,224 @@
+  static std::unique_ptr<DBMain> BuildDBMain(uint16_t network_port, uint16_t messenger_port,
+                                             const std::string &messenger_identity) {
+    std::unordered_map<settings::Param, settings::ParamInfo> param_map;
+    terrier::settings::SettingsManager::ConstructParamMap(param_map);
+
+    auto db_main = terrier::DBMain::Builder()
+                       .SetSettingsParameterMap(std::move(param_map))
+                       .SetUseSettingsManager(true)
+                       .SetUseGC(true)
+                       .SetUseCatalog(true)
+                       .SetUseGCThread(true)
+                       .SetUseTrafficCop(true)
+                       .SetUseStatsStorage(true)
+                       .SetUseLogging(true)
+                       .SetUseNetwork(true)
+                       .SetNetworkPort(network_port)
+                       .SetUseMessenger(true)
+                       .SetMessengerPort(messenger_port)
+                       .SetMessengerIdentity(messenger_identity)
+                       .SetUseExecution(true)
+                       .Build();
+
+    return db_main;
+  }
+};
+
+// NOLINTNEXTLINE
+TEST_F(MessengerTests, BasicReplicationTest) {
+  // TODO(WAN): remove this after demo at meeting.
+  messenger_logger->set_level(spdlog::level::trace);
+
+  uint16_t port_primary = 15721;
+  uint16_t port_replica1 = port_primary + 1;
+  uint16_t port_replica2 = port_primary + 2;
+
+  uint16_t port_messenger_primary = 9022;
+  uint16_t port_messenger_replica1 = port_messenger_primary + 1;
+  uint16_t port_messenger_replica2 = port_messenger_primary + 2;
+
+  // done[3] is shared memory (mmap) so that the forked processes can coordinate on when they are done.
+  // This is done instead of waitpid() because I can't find a way to stop googletest from freaking out on waitpid().
+  // done[0] : primary, done[1] : replica1, done[2] : replica2.
+  bool *done =
+      static_cast<bool *>(mmap(nullptr, 3 * sizeof(bool), PROT_READ | PROT_WRITE, MAP_SHARED | MAP_ANONYMOUS, -1, 0));
+  TERRIER_ASSERT(MAP_FAILED != done, "mmap() failed.");
+
+  done[0] = false;
+  done[1] = false;
+  done[2] = false;
+
+  auto spin_until_done = [done]() {
+    while (!(done[0] && done[1] && done[2])) {
+    }
+  };
+
+  VoidFn primary_fn = [=]() {
+    auto primary = BuildDBMain(port_primary, port_messenger_primary, "primary");
+    primary->GetNetworkLayer()->GetServer()->RunServer();
+
+    while (!done[1]) {
+    }
+
+    MESSENGER_LOG_TRACE("Primary done.");
+    done[0] = true;
+    spin_until_done();
+  };
+
+  VoidFn replica1_fn = [=]() {
+    auto replica1 = BuildDBMain(port_replica1, port_messenger_replica1, "replica1");
+    replica1->GetNetworkLayer()->GetServer()->RunServer();
+
+    // Set up a connection to the primary.
+    auto messenger = replica1->GetMessengerLayer()->GetMessenger();
+    ConnectionDestination dest_primary = Messenger::GetEndpointIPC("primary", port_messenger_primary);
```

Comment:
These tests should still work with replicas on TCP right? Or do we prefer IPC?

Reply:
It should work with TCP. Doesn't matter.